pose_evaluation/metrics/embedding_distance_metric.py

```diff
@@ -1,104 +1,170 @@
 import numpy as np
-
+from sentence_transformers import util as st_util
 from pose_evaluation.metrics.base_embedding_metric import EmbeddingMetric
 
 
+# Useful reference: https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/util.py#L31
+# * Helper functions such as batch_to_device, _convert_to_tensor, _convert_to_batch, _convert_to_batch_tensor
+# * a whole semantic search function, with chunking and top_k
+
+# See also pgvector's C implementation: https://github.com/pgvector/pgvector/blob/master/src/vector.c
+# * cosine_distance: https://github.com/pgvector/pgvector/blob/master/src/vector.c#L658
+# * l2_distance https://github.com/pgvector/pgvector/blob/master/src/vector.c#L566
+
+
 class EmbeddingDistanceMetric(EmbeddingMetric):
-    def __init__(self, kind: Literal["cosine", "l2"] = "cosine", device: Union[torch.device, str] = None):
+    def __init__(
+        self,
+        kind: Literal["cosine", "euclidean", "dot"] = "cosine",
+        device: torch.device | str = None,
+        dtype=torch.float64,
+    ):
         """
         Initialize the embedding distance metric.
 
         Args:
-            kind (Literal["cosine", "l2"]): The type of distance metric.
+            kind (Literal["cosine", "euclidean"]): The type of distance metric.
             device (torch.device | str): The device to use for computation. If None, automatically detects.
         """
         super().__init__(f"EmbeddingDistanceMetric {kind}", higher_is_better=False)
         self.kind = kind
         if device is None:
-            self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+            self.device = torch.device(st_util.get_device_name())
         else:
             self.device = torch.device(device) if isinstance(device, str) else device
 
-    def _to_tensor(self, data: Union[np.ndarray, torch.Tensor]) -> torch.Tensor:
-        """
-        Convert input to a PyTorch tensor if it is a NumPy array.
+        self.dtype = dtype
 
-        Args:
-            data (np.ndarray | torch.Tensor): Input data.
+    def _to_device_tensor(self, data: list | np.ndarray | Tensor, dtype=None) -> Tensor:
+        if dtype is None:
+            dtype = self.dtype
+        return st_util._convert_to_tensor(data).to(device=self.device, dtype=dtype)
 
-        Returns:
-            torch.Tensor: Tensor on the correct device.
-        """
-        if isinstance(data, np.ndarray):
-            data = torch.tensor(data, dtype=torch.float32)
-        return data.to(self.device)
+    def _to_batch_tensor_on_device(self, data: list | np.ndarray | Tensor, dtype=None) -> Tensor:
+        if dtype is None:
+            dtype = self.dtype
+        return st_util._convert_to_batch_tensor(data).to(device=self.device, dtype=dtype)
 
-    def score(self, hypothesis: Union[np.ndarray, torch.Tensor], reference: Union[np.ndarray, torch.Tensor]) -> float:
+    def score(
+        self,
+        hypothesis: list | np.ndarray | Tensor,
```

Comment:
I think maybe it is worth at the top of the file to define a type `Embedding = Union[np.ndarray, torch.Tensor, list]` and then use that type instead of repeating this three

Reply:
Great minds think alike, already did something along those lines, haha